File: Studi/Matdis/responsi.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_MAHASISWA = 100;


struct Mahasiswa {
    string nama;
    string nim;
    int umur;
};

struct DataMahasiswa {
    Mahasiswa mahasiswa[MAX_MAHASISWA];
    int jumlahMahasiswa;
};

void tambahMahasiswa(DataMahasiswa& data) {
    if (data.jumlahMahasiswa == MAX_MAHASISWA) {
        cout << "Data mahasiswa penuh!" << endl;
        return;
    }

    Mahasiswa mhs;
    cout << "Masukkan Nama: ";
    getline(cin, mhs.nama);
    cout << "Masukkan NIM: ";
    getline(cin, mhs.nim);
    cout << "Masukkan Umur: ";
    cin >> mhs.umur;
    cin.ignore();

    data.mahasiswa[data.jumlahMahasiswa++] = mhs;
    cout << "Data mahasiswa berhasil ditambahkan!" << endl;
}

void cariMahasiswa(const DataMahasiswa& data, const string& nim) {
    for (int i = 0; i < data.jumlahMahasiswa; i++) {
        if (data.mahasiswa[i].nim == nim) {
            cout << "Nama   : " << data.mahasiswa[i].nama << endl;
            cout << "NIM    : " << data.mahasiswa[i].nim << endl;
            cout << "Umur   : " << data.mahasiswa[i].umur << endl;
            return;
        }
    }
    cout << "Mahasiswa dengan NIM " << nim << " tidak ditemukan." << endl;
}

void hapusMahasiswa(DataMahasiswa& data, const string& nim) {
    for (int i = 0; i < data.jumlahMahasiswa; i++) {
        if (data.mahasiswa[i].nim == nim) {
            for (int j = i; j < data.jumlahMahasiswa - 1; j++) {
                data.mahasiswa[j] = data.mahasiswa[j + 1];
            }
            data.jumlahMahasiswa--;
            cout << "Data mahasiswa dengan NIM " << nim << " berhasil dihapus!" << endl;
            return;
        }
    }
    cout << "Mahasiswa dengan NIM " << nim << " tidak ditemukan." << endl;
}

void tampilkanMahasiswa(const DataMahasiswa& data) {
    if (data.jumlahMahasiswa == 0) {
        cout << "Data mahasiswa kosong." << endl;
        return;
    }

    cout << "Daftar Mahasiswa:" << endl;
    for (int i = 0; i < data.jumlahMahasiswa; i++) {
        cout << "Nama   : " << data.mahasiswa[i].nama << endl;
        cout << "NIM    : " << data.mahasiswa[i].nim << endl;
        cout << "Umur   : " << data.mahasiswa[i].umur << endl;
        cout << endl;
    }
}

int main() {
    cout << "";
    DataMahasiswa data;
    data.jumlahMahasiswa = 0;

    int pilihan;
    string nim;

    do {
        cout << "Menu:" << endl;
        cout << "1. Tambah Mahasiswa" << endl;
        cout << "2. Cari Mahasiswa" << endl;
        cout << "3. Hapus Mahasiswa" << endl;
        cout << "4. Tampilkan Mahasiswa" << endl;
        cout << "0. Keluar" << endl;
        cout << "Pilih menu: ";
        cin >> pilihan;
        cin.ignore();

        switch (pilihan) {
            case 1:
                tambahMahasiswa(data);
                break;
            case 2:
                cout << "Masukkan NIM: ";
                getline(cin, nim);
                cariMahasiswa(data, nim);
                break;
            case 3:
                cout << "Masukkan NIM: ";
                getline(cin, nim);
                hapusMahasiswa(data, nim);
                break;
            case 4:
                tampilkanMahasiswa(data);
                break;
            case 0:
                cout << "Terima kasih!" << endl;
                break;
            default:
                cout << "Pilihan tidak valid!" << endl;
                break;
        }
        cout << endl;

    } while (pilihan != 0);

    return 0;
}

File: Studi/Matdis/Responsi.py
# Author : Muhamad Fadhli Akbar - 2200018197
# Judul  : Pengelolaan Data Mahasiswa menggunakan Relasi
# Slot   : Sabtu 07.30


import os


class Mahasiswa:
    def __init__(self, nim, nama, jurusan):
        self.nim = nim
        self.nama = nama
        self.jurusan = jurusan


class DatabaseMahasiswa:
    def __init__(self):
        self.mahasiswa = []

    def tambah_mahasiswa(self, nim, nama, jurusan):
        mhs = Mahasiswa(nim, nama, jurusan)
        self.mahasiswa.append(mhs)
        print("Data mahasiswa berhasil ditambahkan.")

    def cari_mahasiswa(self, nim):
        for mhs in self.mahasiswa:
            if mhs.nim == nim:
                return mhs
        return None

    def hapus_mahasiswa(self, nim):
        mhs = self.cari_mahasiswa(nim)
        if mhs:
            self.mahasiswa.remove(mhs)
            print("Data mahasiswa berhasil dihapus.")
        else:
            print("Data mahasiswa tidak ditemukan.")

    def tampilkan_semua(self):
        if len(self.mahasiswa) > 0:
            print("Daftar Mahasiswa:")
            for mhs in self.mahasiswa:
                print(
                    f"NIM: {mhs.nim}, Nama: {mhs.nama}, Jurusan: {mhs.jurusan}")
        else:
            print("Belum ada data mahasiswa.")


def clear_screen():
    # Sesuaikan sesuai dengan sistem operasi yang Anda gunakan
    if os.name == 'posix':  # Untuk sistem operasi berbasis Unix (Linux, macOS)
        _ = os.system('clear')
    else:  # Untuk sistem operasi Windows
        _ = os.system('cls')


db = DatabaseMahasiswa()

while True:
    print("----------------------------------")
    print("Program Pengelolaan Data Mahasiswa")
    print("----------------------------------")
    print("1. Tambah Mahasiswa")
    print("2. Cari Mahasiswa")
    print("3. Hapus Mahasiswa")
    print("4. Tampilkan Semua Mahasiswa")
    print("5. Keluar")

    pilihan = input("Masukkan Pilihan anda : ")
    clear_screen()  # Clear screen setelah pengguna memilih opsi

    if pilihan == "1":
        nim = input("Masukkan NIM: ")
        nama = input("Masukkan Nama: ")
        jurusan = input("Masukkan Jurusan: ")
        db.tambah_mahasiswa(nim, nama, jurusan)
    elif pilihan == "2":
        nim = input("Masukkan NIM: ")
        mhs = db.cari_mahasiswa(nim)
        if mhs:
            print(
                f"Mahasiswa ditemukan.\nNIM: {mhs.nim}, Nama: {mhs.nama}, Jurusan: {mhs.jurusan}")
        else:
            print("Mahasiswa tidak ditemukan.")
    elif pilihan == "3":
        nim = input("Masukkan NIM: ")
        db.hapus_mahasiswa(nim)
    elif pilihan == "4":
        db.tampilkan_semua()
    elif pilihan == "5":
        print("Program selesai.")
        break
    else:
        print("Pilihan tidak valid. Silakan masukkan angka 1-5.")

    input("Tekan Enter untuk melanjutkan...")
    clear_screen()  # Clear screen sebelum menampilkan menu berikutnya
